src/common/browser-adapters/browser-adapter-event-manager.ts

```diff
@@ -0,0 +1,132 @@
+    constructor(
+        private promiseFactory?: PromiseFactory,
+        private setTimeoutManager?: SetTimeoutManager,
+    ) {}
+
+    public registerEventToApplicationListener = (
+        eventType: string,
+        callback: ApplicationListener,
+    ): void => {
+        if (this.eventsToApplicationListenersMapping[eventType]) {
+            this.eventsToApplicationListenersMapping[eventType].push(callback);
+        } else {
+            this.eventsToApplicationListenersMapping[eventType] = [callback];
+        }
+        this.processDeferredEvents();
+    };
+
+    public registerAdapterListenerForEvent = (
+        event: Events.Event<any>,
+        eventType: string,
+    ): void => {
+        event.addListener(this.adapterListener(eventType));
+    };
+
+    private processDeferredEvents(): void {
+        this.deferredEvents = this.deferredEvents.filter(
+            event => this.processEvent(event.eventType, event.eventArgs, true) === false,
+        );
+    }
+
+    private async forwardEventToApplicationListener(
+        listener: ApplicationListener,
+        eventArgs: any[],
+    ): Promise<any> {
+        const result = listener(...eventArgs);
+        if (!!result && typeof result.then === 'function') {
+            if (this.promiseFactory) {
+                //wrapping application listener promise responses in a 4-minute promise race
+                //prevents the service worker going idle before a response is sent
+                await this.promiseFactory.timeout(result, 240000);
+            }
+        } else {
+            // it is possible that this is the result of a fire and forget listener
+            // wrap in 2-minute timeout to ensure it completes during service worker lifetime
+            if (this.setTimeoutManager) {
+                return this.setTimeoutManager.setTimeout(
+                    () => Promise.resolve(result),
+                    120000,
+                    `wrapper-timeout-${Date.now()}`,
+                );
+            }
+        }
+        return result;
+    }
+
+    private unregisterAdapterListener(event: Events.Event<any>, eventType: string) {
+        event.removeListener(this.adapterListener(eventType));
+    }
+
+    private unregisterApplicationListenerForEvent(
+        eventType: string,
+        listener: ApplicationListener,
+    ) {
+        const allListeners = this.eventsToApplicationListenersMapping[eventType];
+        const applicationListenerIndex = allListeners.findIndex(listener);
+        if (applicationListenerIndex < 0) {
+            console.error('no application listeners found for ', eventType);
+        } else {
+            if (allListeners.length === 1) {
+                delete this.eventsToApplicationListenersMapping[eventType];
+            } else {
+                this.eventsToApplicationListenersMapping[eventType] =
+                    allListeners.splice(applicationListenerIndex);
+            }
+        }
+    }
+
+    public removeListener(
```

Comment:
Is there any test coverage for this method?

Reply:
There isn't right now because it isn't used anywhere. None of the unregister or remove methods are. I'd plan on adding tests once we determine what graceful teardown looks like.

Reply:
Makes sense, yeah I'd prefer to omit them for now just to make sure we don't add them now then forget to add tests later

Reply:
I ended up writing tests for them after remembering where they are used later. We'll need them for my next PR anyhow.